README.md
# nmacs
C++ linux branch of an ncurses text editor 

TODO: add screenshots, use cases, etc.


Changelog
Sat Dec 06, 2014 22:46:32
	- Got a working menu class.  Still need to integrate with buffer class.
	- the first step is done:  hopefully changed all buffer interactions to a subwin instead of stdscr
Fri Dec 05, 2014 03:19:25
	- Got file reads working from file mode
	- Now I just need to get file writes to work
	- next steps after that are writing to a new file
	
Fri Dec 05, 2014 00:30:50
	- Still can't open files from file editor
	- Could define keymaps with function pointers: http://www.cplusplus.com/forum/general/14848/
	
Thu Dec 04, 2014 07:14:02
	- added some polymorphism.
	- still can't open files.
	- still can't switch buffers
	- need to add keybindings for fileEditor mode

Thu Dec 04, 2014 04:00:23
	- File editors still can't open files for some reason
	- Text editors can open files, but the key presses are screwy.

	- I'm trying to move keypresses out of the buffer base class
	- Just trying to move them into world.
	- tab keys break things
Sat Nov 18, 2017 
        - updates.  Trying to find a way to work on this more regularly


menu.h
/*
  TODO: Review, document, etc.
 */

#ifndef MENU_H
#define MENU_H

#include <ncurses.h>
//#include <stdlib.h>
#include <cstdlib>
#include <vector>
#include <string>

#define ENTER 10
#define ESCAPE 27

using std::vector;
using std::string;

struct dropDownMenu{
  string name;
  int xStart; // x-coordinate to draw from
  vector<string> subItem; // subitems
  vector<WINDOW*> window;
  // vector of function pointers? TODO
};


class menuHeader{
public:
  // constructor
  menuHeader();  // end constructor function
  ~menuHeader();
  // member variables
  int key;
  WINDOW *menubar; // menubar contains subwins?
  WINDOW *messagebar; 
  vector<dropDownMenu> subMenu; // container for dropDownMenus

  // member functions
  
  int mod(int a, int b) { 
    int r = a % b;
    return r < 0 ? r + b : r;
  }
  
  string getKeys();
  string getKeys(int key);
  string handleSubMenu(int num);
  void setStartCoordinates();
  // draws all submenu names in class horizontally across screen
  void drawMenuBar();
  // recurse over vector of names
  int getSubMenuNameLengths(int n, int sum);
  void drawSubMenu(int number);  
  void deleteSubMenu(int number);  
  string scrollSubMenu(int number);  
}; // end menuHeader class

#endif //MENU_H


buffer.c
/*
  Sat Dec 20, 2014 22:03:02 
  Really striving for maximum simplicity here. 

  Just an array of strings that can be printed on ncurses to test scrolling.

  This version uses a plain GArray
 */
#include <ncurses.h>
#include <glib.h>
//#include <stdlib.h>

#ifndef min
#define min(a,b) (((a) < (b)) ? (a) : (b))
#endif

#define ENTER 10
#define TAB 9
#define ESCAPE 27


GString* g_substring (GString *str, int index, int len)
{
  GString* returnVal = g_string_new(NULL);
  if (len <= str->len)
    for (int i = index; i < len; i++)
      {
	gchar j = str->str[i];
	g_string_append_printf(returnVal, "%c",j);
      }
    //  return g_string_new_len (str->str, min (str->len - index, len));
  return returnVal;
}


// // a line has a number and text
// typedef struct Line {  
//   int number;
//   // string text;
//   //GArray* text;
//   GString* text;
// } line;

// typedef struct Location {
//   int x;
//   int y;
// } location;

// print2vec(const char cstring[]){
  
// }

void stats(GArray* lines,
	   int currentLine,
	   int startLine,
	   int finishLine)
  {
    wattron(stdscr,A_REVERSE);
    int ty,tx;
    getyx(stdscr,ty,tx); // store point
    int cy = ty;
    int cx = tx;
    int ymax,xmax;
    getmaxyx(stdscr,ymax,xmax);
    
    wmove(stdscr,ymax-1,0); 
    int lineNumber = currentLine + 1;

    //wmove(stdscr,finishLine,0); 
    wclrtoeol(stdscr); // clear line
    wmove(stdscr,ymax-1,0);
    wattron(stdscr,A_BOLD);
    // wprintw(stdscr,"%s: ",name.c_str());
    wattroff(stdscr,A_BOLD);
    wprintw(stdscr,"cx:%d,cy:%d,sl:%d,cl:%d,fl:%d,xmax:%d,ymax:%d,ln:%d,lines:%d",
	    cx,cy,startLine,currentLine,finishLine,xmax,ymax,lineNumber,lines->len);
    wrefresh(stdscr);
    wmove(stdscr,ty,tx); // restore point
    wattroff(stdscr,A_REVERSE);
  }


void displayLines(GArray* lines,
		  int startLine,
		  int finishLine){
  int xmax,ymax,ty,tx;
  getmaxyx(stdscr,ymax,xmax);
  getyx(stdscr,ty,tx);
  clear();
  //refresh();
  for (int i = startLine; i < finishLine; i++) // get ith line
    { 
      wmove(stdscr,startLine-i,0);
      // g_ptr_array_index(lines,i) contains ith line
      // need to extract from it the jth char
      // store ith line as a new GString
      GString* line;
      line = g_array_index(lines,GString*,i);
      // print jth char of ith line
      for (int j = 0; j < min(xmax,(int)line->len); j++) 
	//for (int j = 0; j < line->len; j++)
	wprintw(stdscr,"%c",line->str[j]);
    }
  wmove(stdscr,ty,tx);
  refresh();
}

int getStartLine(GArray* lines,
		  int currentLine)
{
  int cy,cx;
  getyx(stdscr,cy,cx);
  int minBoundary = currentLine - cy;
  // startLine must be >= 0 
  return (minBoundary >= 0) ? minBoundary : 0;
}

int getFinishLine(GArray* lines,
		  int currentLine)
{
  int cy,cx, xmax,ymax;
  getyx(stdscr,cy,cx);
  getmaxyx(stdscr,ymax,xmax);
  int maxBoundary = currentLine + ymax - cy - 1;
  return (maxBoundary < lines->len) ? maxBoundary : lines->len;
}


void scrollBufferREPL(GArray* lines){
  int xmax,ymax,cx,cy;
  getmaxyx(stdscr,ymax,xmax);
  int key;

  int startLine = 0;
  int currentLine = 0;
  int finishLine = ymax-1;

  displayLines(lines,startLine,finishLine);

  wmove(stdscr,0,0);


  bool editorRunning = true;
  

  while (editorRunning)
    {
      getmaxyx(stdscr,ymax,xmax);
      getyx(stdscr,cy,cx);
      key = getch();  
      switch (key)
	{
	case KEY_UP:
	  {
	    if (cy-1 >= 0) // don't scroll display 
	      {
	    	wmove(stdscr,cy-1,cx);
	    	currentLine--;
	      }
	    else // scroll display
	      {
	    	if (currentLine-1 >= 0)
	    	  {
	    	    wmove(stdscr,cy-1,cx);
	    	    currentLine--;
	    	    //startLine = getStartLine(lines,currentLine);
	    	    //finishLine = getFinishLine(lines,currentLine);
	    	    startLine--;
	    	    finishLine--;
	    	  }
	      }
	    //printw("key_up pressed");
	    break;
	  } // end key_up
	case KEY_DOWN:
	  {
	    //if (currentLine+1 < ymax-1)
	    if (cy+1 < ymax-1) // don't scroll display
	      {
		if (cy+1 < lines->len)
		  {
		    wmove(stdscr,cy+1,cx);
		    currentLine++;
		    //refresh();
		  }
		// wmove(stdscr,cy+1,cx);
	      }
	    else // scroll display
	      {
		if (currentLine+1 < lines->len)
		  {
		    currentLine++;
		    startLine = getStartLine(lines,currentLine);
		    finishLine = getFinishLine(lines,currentLine);
		    //startLine++;
		    //finishLine++;
		  }
	      }
	    //displayLines(lines,startLine,finishLine);
	    break;
	  } // end key_down
	case KEY_RIGHT: 
	  {
	    GString* line;
	    line = g_array_index(lines,GString*,currentLine);
	    if (cx+1 < line->len)
	      wmove(stdscr,cy,cx+1);
	    else // duplicate key_down functionality
	      {
		// if (cy+1 < ymax-1)
		//   if (currentLine+1 < lines->len)
		//     {
		//       wmove(stdscr,cy+1,0);
		//       currentLine++;
		//     }
		
		// need to duplicate key_down functionality.
		// key_down(lines,currentLine,cy,cx);
		// this doesn't warrant a broken out function just yet
		
		/*
		  Sun Dec 21, 2014 21:05:22
		  Actually, this is a slightly different function that
		  wouldn't work correctly by simply calling key_down
		  if it existed.
		 */

		if (cy+1 < ymax-1) // don't scroll display
		  {
		    if (cy+1 < lines->len)
		      {
			wmove(stdscr,cy+1,0);
			currentLine++;
			//refresh();
		      }
		    wmove(stdscr,cy+1,0);
		  }
		else // scroll display
		  {
		    if (currentLine+1 < lines->len)
		      {
			currentLine++;
			startLine = getStartLine(lines,currentLine);
			finishLine = getFinishLine(lines,currentLine);
			wmove(stdscr,cy,0); // put cursor at beginning of line
			//startLine++;
			//finishLine++;
		      }
		  }

	      } // 
	    break;
	  } // end key_right
	case KEY_LEFT: 
	  {
	    if (cx-1 >= 0) // stay on same line
	      wmove(stdscr,cy,cx-1);
	    else // move to prev line
	      {
		if (currentLine-1 >= 0)
		  {
		    currentLine--;
		    GString* line;
		    line = g_array_index(lines,GString*,currentLine);
		    wmove(stdscr,cy-1,
			  line->len-1
			  );
		  }
	      }
	    break;
	  } // end key_left
	case KEY_BACKSPACE: 
	  {
	    getyx(stdscr,cy,cx);
	    GString *before;
	    GString *after;
	    GString *current;

	    if (cx-1 >= 0) // stay on same line
	      {
		// // split current line text into two strings
		current = g_array_index(lines,GString*,currentLine);
		before = g_substring(current,0,cx-1); 
		after = g_substring(current,cx,current->len);
		
		// // push character into string "before" point
		///g_string_append_printf(before, "%c",(char)key);
		
		// // concat "before" and "after" strings
		g_string_assign(current, before->str); // push before into current
		g_string_append_printf(current,"%s",after->str); 

		// // restore line in structure
		g_array_insert_vals(lines,currentLine,&current,0);
		// note the ampersand for "current".  I'm not sure of these details 
		
		// // update cursor
		wmove(stdscr,cy,cx-1);

	      }
	    else // move to prev line
	      {
		if (currentLine > 0)
		  {
		    currentLine--;
		    current = g_array_index(lines,GString*,currentLine);
		    before = g_substring(current,0,current->len-1); 
		    after = g_array_index(lines,GString*,currentLine+1);
		    g_string_assign(current,before->str);
		    g_string_append_printf(current,"%s",after->str);
		    wmove(stdscr,cy-1,before->len);
		    g_array_remove_index(lines,currentLine+1);
		  }
	      }
	    refresh();
	    startLine = getStartLine(lines,currentLine);
	    finishLine = getFinishLine(lines,currentLine);

	    break;
	  } // end key_backspace
	  
	  //case KEY_ENTER:
	case ENTER:
	  {
	    /*
	      Tue Dec 30, 2014 12:24:45
	      Note that g_array_insert_vals()  needs to use the ampersand &
	      to address the value to insert for some reason. 

	      Took me hours to realize this.
	     */
	    if (cx == 0)
	      {
		GString *emptyLine;
		emptyLine = g_string_new("\n");
		g_array_insert_vals(lines,(guint)currentLine,&emptyLine,1);
		currentLine++;      
		wmove(stdscr,cy+1,0);
		//startLine = getStartLine(lines,currentLine);
		//finishLine = getFinishLine(lines,currentLine);
	      }
	    else
	      {
		GString* before;
		GString* after;
		GString* current;
		current = g_array_index(lines,GString*,currentLine);
		before = g_substring(current,0,cx); 
		g_string_append(before,"\n");
		after = g_substring(current,cx,current->len); 

		g_array_remove_index(lines,currentLine);
		g_array_insert_vals(lines,currentLine,&before,1);
		g_array_insert_vals(lines,currentLine+1,&after,1);  // should get pushed to next line in next step



		currentLine++;
		wmove(stdscr,cy+1,0);
		startLine = getStartLine(lines,currentLine);
		finishLine = getFinishLine(lines,currentLine);
	      }
	    break;
	  }
	default:  // line gap
	  {
	    getyx(stdscr,cy,cx);
	    GString *before;
	    GString *after;
	    GString *current;
	    // current = lines[currentLine].text;
	    current = g_array_index(lines,GString*,currentLine);
	    // // split current line text into two strings
	    // // 1 before cursor, 1 after cursor

	    // before = current.substr(0,point.x); // includes current cursor
	    // before = g_strndup(current,cx); // includes current cursor
	    before = g_substring(current,0,cx); // includes current cursor

	    // after = current.substr(point.x,current.length());
	    // after = g_strndup(current+cx,current->len);
	    after = g_substring(current,cx,current->len);

	    // // push character into string "before" point
	    // before.push_back(thisChar);
	    // g_string_append(before, (gchar) key);
	    g_string_append_printf(before, "%c",(char)key);

	    // // concat "before" and "after" strings
	    // current = before + after;
	    g_string_assign(current, before->str); // push before into current
	    g_string_append_printf(current,"%s",after->str); 


	    // // update cursor
	    // wmove(bufSubWin,point.y,point.x+1);
	    wmove(stdscr,cy,cx+1);

	    // // restore line in structure
	    // lines[currentLine].text = current;
	    // g_array_insert_val(lines,currentLine,&current);
	    g_array_insert_vals(lines,currentLine,current,0);

	    // clean up
	    // g_string_free(before,TRUE);
	    // g_string_free(after,TRUE);
	    // g_string_free(current,TRUE);

	    break;
	  }
	} // end switch
      displayLines(lines,startLine,finishLine);
      stats(lines,
	    currentLine,
	    startLine,
	    finishLine);

    } // end REPL
}

int main()
{
  // init curses
  // -----------
  initscr();
  cbreak();
  curs_set(2);
  noecho();
  keypad(stdscr,true);
  // end curses init

  // setup structures
  // ------------------
  //GString line; 
  int ymax,xmax;
  getmaxyx(stdscr,ymax,xmax);
  GArray* lines; // array of lines
  lines = g_array_new(FALSE,FALSE,sizeof(GString *));
  // can find array size with lines->len which will return a guint
  // a guint is an unsigned int



  // add text to lines
  // -------------------
  GString* ar;
  for (int i = 0; i < 100; i++)
    {
      ar = g_string_new(NULL); // initialize string
      // push text to string
      g_string_append_printf(ar, 
			     "%d - lots and lots of lines flowing down the terminal\n",i);
      // push string into array
      g_array_append_val(lines, ar);
    }


  //displayLines(lines);
  // REPL starts here
  scrollBufferREPL(lines);

  endwin();
}


buffer.h
/*
  Fri Dec 05, 2014 01:47:09
  This is a simplification.  

  There should only be one class here that handles everything. 
  
  Likewise, there should only be one editing mode.  
  TODO: code review
 */
#ifndef BUFFER_H
#define BUFFER_H

#include <ncurses.h>
#include <vector>
#include <string>
#include <cstring>
#include <stdio.h>
#include <algorithm> // min()
#include <fstream>
#include "boost/filesystem.hpp"
#include "boost/filesystem/path.hpp"
#include "boost/filesystem/operations.hpp"
#include <stdlib.h> // getline
#include <sstream>
#include <unistd.h>  // usleep

using std::vector;
using std::string;
using std::min;
using std::stringstream;
using std::endl;

namespace fs = boost::filesystem;

#define ENTER 10
#define TAB 9
#define ESCAPE 27


// a line has a number and text
struct line{  
  int number;
  string text;
};

struct location{
  int x;
  int y;
};

// a buffer has lines, but it also has a display 
class buffer{
public:
  // variables
  WINDOW* bufSubWin;
  vector<line> lines;
  string mode;
  int xmax; // maximum x position determined by screen size
  int ymax; // maximum y position determined by screen size
  location point; // current x,y location on screen
  location mark; // location to cut/paste from
  //int cx; // current x position on screen
  //int cy; // current y position on screen
  int startLine; // line at top of screen
  int finishLine; // line at bottom of screen
  int currentLine; // line that cursor is on
  int lineNumber; // always should be currentLine + 1
  string name; // name of buffer
  string CurrentFileName; // filename for buffer
  bool editorIsRunning;
  int keyPress;
  string pathName;
  // functions
  virtual void init(); // set necessary variables in object
  virtual void stats(); // show lines in buffer, current position
  virtual void update(); // get screen dimensions and line values
  virtual void getStartFinish(); // find startLine and finishLine for screen rendering
  virtual void updateLineNumbers(); // renumber lines appropriately
  virtual void print2vec(const char cstring[]); // pushes text into vector of lines
  virtual void displayLines(); // renders lines to fit on screen
  virtual void setName(const string&); // set name of buffer
  string getName(); // return name of buffer

  //virtual void listFiles();
  virtual void listFiles(string&);
  virtual void openFile(string);
  virtual void writeFile(string);
  virtual void getCwd();

  void getUserInput();
  void getUserInput(int keyPress);
  
  void key_up();
  void key_down();
  void key_left();
  void key_right();
  void key_home();
  void key_end();
  void key_pgUp();
  void key_pgDown();
  void key_backspace();
  void key_enter_textMode();
  void key_enter_fileMode();
  void lineGap(int);

//void openFile(string&);
//void writeFile(string&);

};

#endif //BUFFER_H


main.cpp
#include "buffer.h"
#include "menu.h"

// TODO:  make this into a factory pattern - get rid of if/else
class bufferMenu{
public:
  int key;
  buffer editor;
  menuHeader header;
  void handleCallBack(string callBack){
    if (callBack == "New"){
      ;
    }
    if (callBack == "Open"){
      editor.openFile(string("./"));
    }
    if (callBack == "Save"){
      if (editor.CurrentFileName != " ")
	editor.writeFile(editor.CurrentFileName);
      else
	// prompt for filename
	;
    }
    if (callBack == "Save As"){
      // open new window to prompt for fileName
      ;
    }
    if (callBack == "List All Buffers"){
      ;
    }
    if (callBack == "Cut"){
      ;
    }
    if (callBack == "Copy"){
      ; 
    }
    if (callBack == "Paste"){
      ;
    }
    if (callBack == "Select All"){
      ;
    }
    if (callBack == "Find"){
      ;
    }
    if (callBack == "Preferences"){
      ;
    }
    if (callBack == "Next Buffer"){
      ;
    }
    if (callBack == "Prev Buffer"){
      ;
    }
    if (callBack == "Life"){
      // system("./gameOfLife");
    }
    if (callBack == "Pong"){
      ;
    }
    if (callBack == "Space Invaders"){
      ;
    }

  }
  void getUserInput(){
    while (1)
      {
	key = getch();
	//string callBack;
	int ty,tx;
	getyx(editor.bufSubWin,ty,tx);
	
	header.drawMenuBar();
	wmove(editor.bufSubWin,ty,tx);
	
	//werase(nmacs.header.messagebar);
	//wrefresh(nmacs.header.messagebar);
	if (key==KEY_F(1)) { handleCallBack( header.handleSubMenu(0) );   } 
	else if (key==KEY_F(2)) { handleCallBack( header.handleSubMenu(1) );     }
	else if (key==KEY_F(3)) { handleCallBack( header.handleSubMenu(2) );     }
	else if (key==KEY_F(4)) { handleCallBack( header.handleSubMenu(3) );     }
	else { editor.getUserInput(key); }
	//
	curs_set(2);
      }
  
  }
}; // end bufferMenu class

int main()
{

  initscr();
  cbreak();
  curs_set(2);
  noecho();

  // this section is iffy
  start_color();
  init_pair(1,COLOR_GREEN,COLOR_BLACK);
  init_pair(2,COLOR_BLACK,COLOR_GREEN);
  init_pair(3,COLOR_BLUE,COLOR_GREEN);


  keypad(stdscr,true);

  bufferMenu nmacs;
  //menuHeader nmacs;
  int key = 0;

  nmacs.editor.init();
  nmacs.editor.update();

  nmacs.editor.mode = "file";
  string filename = string("Makefile");
  nmacs.editor.openFile(filename);
  nmacs.header.drawMenuBar();
  nmacs.getUserInput();



  

  
  // getch();
  endwin();
  return 0;
}


menu.cpp
/*

  Sat Dec 06, 2014 21:32:57 fixed segfaults, fixed menu wrapping

  Sat Dec 06, 2014 15:59:32   packed everything into a class.  Doesn't run. 
  19:34:26 Now it runs.

  Sat Dec 06, 2014 12:30:42
  14:00:05 : segfaults on scroll_menu(menuBar).   For some reason,
  this isn't getting the correct size of the vector<string> subItem.

  I suspect this is due to pointer errors.

  Fri Dec 05, 2014 14:00:12
  (browse-url "http://www.linuxfocus.org/English/March2002/article233.shtml")
  
  16:03:00 example-02.cpp
  Two hours later. This works using vectors instead of c_arrays.

  Could still clean it up quite a bit.  Maybe put everything into
  classes instead of having a main() driver.

 * todo [3/5]
   - [X] wrap everything into a generic class somehow
   - [X] fix key_left key_right arrow problems
   - [X] fix delete window problem (doesn't delete)
   - [ ] functions in menu don't actually execute yet
   - [ ] merge with buffer.cpp somehow to execute functions
   - [ ] 2017/11/18 code review.  Why is this even here?


 */

#include "menu.h"

menuHeader::menuHeader(){
    printf("starting menuHeader");

    dropDownMenu fileMenu;
    dropDownMenu editMenu;
    dropDownMenu bufferMenu;
    dropDownMenu gameMenu;


    // pack subitems into file menu
    fileMenu.name =  string("(F1): File");
    fileMenu.subItem.push_back(string(""));
    fileMenu.subItem.push_back(string("New"));
    fileMenu.subItem.push_back(string("Open"));
    fileMenu.subItem.push_back(string("Save"));
    fileMenu.subItem.push_back(string("Save As"));
    fileMenu.subItem.push_back(string("List All Buffers"));
    fileMenu.subItem.push_back(string("Close"));
    fileMenu.subItem.push_back(string("Quit"));

    // pack subitems into edit menu    
    editMenu.name =  string("(F2): Edit");
    editMenu.subItem.push_back(string(""));
    editMenu.subItem.push_back(string("Cut"));
    editMenu.subItem.push_back(string("Copy"));
    editMenu.subItem.push_back(string("Paste"));
    editMenu.subItem.push_back(string("Select All"));
    editMenu.subItem.push_back(string("Find"));
    editMenu.subItem.push_back(string("Preferences"));

    // pack subitems into buffer menu    
    bufferMenu.name =  string("(F3): Buffer");
    bufferMenu.subItem.push_back(string(""));
    bufferMenu.subItem.push_back(string("List All"));
    bufferMenu.subItem.push_back(string("Next Buffer"));
    bufferMenu.subItem.push_back(string("Prev Buffer"));

    // pack subitems into game menu    
    gameMenu.name =  string("(F4): Game");
    gameMenu.subItem.push_back(string(""));
    gameMenu.subItem.push_back(string("Life"));
    gameMenu.subItem.push_back(string("Pong"));
    gameMenu.subItem.push_back(string("SpaceInvaders"));

    // pack submenus into header menu
    subMenu.push_back(fileMenu);
    subMenu.push_back(editMenu);
    subMenu.push_back(bufferMenu);
    subMenu.push_back(gameMenu);
    
    setStartCoordinates();

    bkgd(COLOR_PAIR(1));
    menubar=subwin(stdscr,1,80,0,0);
    //header.win=subwin(stdscr,1,80,0,0);
    messagebar=subwin(stdscr,1,79,23,1);
    // draw_menubar(menubar);
    drawMenuBar();
    // move(2,1);
    // printw("Press F1 or F2 to open the menus. ");
    // printw("ESC quits.");
    refresh();

    //getKeys();
    // do {
    //   //WINDOW **menu_items;
    //   //vector<WINDOW*> menu_items;
    //   key=getch();
    //   werase(messagebar);
    //   wrefresh(messagebar);
    //   if (key==KEY_F(1)) { handleSubMenu(0);  } 
    //   else if (key==KEY_F(2)) { handleSubMenu(1);     }
    //   else if (key==KEY_F(3)) { handleSubMenu(2);     }
    //   else if (key==KEY_F(4)) { handleSubMenu(3);     }

    // } while (key!=ESCAPE);
    
  } // end constructor function
 
menuHeader::~menuHeader() {
    delwin(menubar);
    delwin(messagebar);
    printf("ending menuHeader");
  } // end destructor function
  
string menuHeader::getKeys(){
  do {
    key=getch();
    // werase(messagebar);
    // wrefresh(messagebar);
    // if (key==KEY_F(1)) { handleSubMenu(0);  } 
    // else if (key==KEY_F(2)) { handleSubMenu(1);     }
    // else if (key==KEY_F(3)) { handleSubMenu(2);     }
    // else if (key==KEY_F(4)) { handleSubMenu(3);     }
    getKeys(key);
  } while (key!=ESCAPE);
} // end getKeys()


string menuHeader::getKeys(int key){
  //  do {
  //key=getch();
    werase(messagebar);
    wrefresh(messagebar);
    if (key==KEY_F(1)) { handleSubMenu(0);  } 
    else if (key==KEY_F(2)) { handleSubMenu(1);     }
    else if (key==KEY_F(3)) { handleSubMenu(2);     }
    else if (key==KEY_F(4)) { handleSubMenu(3);     }
    
    //  } while (key!=ESCAPE);
} // end getKeys()


string menuHeader::handleSubMenu(int num){
  int ty,tx;
  getyx(stdscr,ty,tx);
  // int selected_item;
  string selected_item;
  drawSubMenu(num);
  // selected_item=scroll_menu(fileMenu,8,0);
  selected_item=scrollSubMenu(num);
  //delete_menu(fileMenu,9);
  deleteSubMenu(num);
  // if (selected_item<0)
  //   wprintw(messagebar,"You haven't selected any item.");
  // else
  
  /*
    wprintw(messagebar,
    //"You have selected menu item %d.",selected_item+1);
    //"You have selected %s.",fileMenu.name[selected_item+1].c_str());
    //"You have selected %s.",subMenu[num].subItem[selected_item+1].c_str());
    selected_item.c_str());
  */
  
  touchwin(stdscr);
  refresh();
  move(ty,tx);
  return selected_item;
} // end handleSubMenu()

  void menuHeader::setStartCoordinates(){
    if (subMenu.size() >= 0){
      subMenu[0].xStart = 0; // or padded number
      if (subMenu.size() >= 1){
	for (int i = 1; i < subMenu.size(); i++)
	  // subMenu[i].xStart = subMenu[i-1].name.length()+1;
	  // NO!  it's the sum of lengths > 0!
	  subMenu[i].xStart = getSubMenuNameLengths(i,0);
      }
    }
  } // end setStartCoodinates()

  // draws all submenu names in class horizontally across screen
  void menuHeader::drawMenuBar(){
    string itemName;
    int tx,ty;
    getyx(stdscr,ty,tx);
    wbkgd(menubar,COLOR_PAIR(2));
    wmove(menubar,0,0);
    for (int i = 0; i < subMenu.size(); i++){

      // itemName = subMenu[i].name;
      // //itemName = "(F1): File";
      // waddstr(menubar,itemName.c_str());
      //wmove(menubar,0,itemName.length());
      wprintw(menubar,"%s ",subMenu[i].name.c_str());
      // if (i+1 < subMenu.size()){
      // 	// 	subMenu[i+1].xStart = getSubMenuNameLengths();
      // 	getyx(menubar,NULL,subMenu[i+1].xStart);
      // 	subMenu[i+1].xStart++;
      // }
    }
    wrefresh(menubar);
    move(ty,tx);
    curs_set(2);
  } // end drawMenuBar()

  // recurse over vector of names
  int menuHeader::getSubMenuNameLengths(int n, int sum){
    //static int sum = 0;
    // int nameLength;
    if (n < 0)
      return sum;
    else 
      sum += subMenu[n-1].name.length();
    // nameLength = subMenu[number-1].name.length();
    return getSubMenuNameLengths(--n,sum);
  }

  void menuHeader::drawSubMenu(int number) {
    int ty,tx;
    getyx(stdscr,ty,tx);
    dropDownMenu& thisMenu = subMenu[number];
    if (thisMenu.window.size() > 0)
      thisMenu.window.clear(); 
    thisMenu.window.push_back(newwin(thisMenu.name.size()+1,
				     19,
				     1,
				     thisMenu.xStart));
    wbkgd(thisMenu.window[0],COLOR_PAIR(2));
    box(thisMenu.window[0],ACS_VLINE,ACS_HLINE);
    // WINDOW *subwin(WINDOW *orig, int nlines, int ncols, int begin_y, int begin_x);
    //  for (int i=1; i<9; i++)
    for (int i=1; i<thisMenu.subItem.size(); i++)
      //thisMenu.window.push_back(subwin(thisMenu.window[0],1,17,i+1,thisMenu.xStart)); 
      thisMenu.window.push_back(subwin(thisMenu.window[0],
				       1,
				       17,
				       i+1,
				       thisMenu.xStart+1)); 
    
    // print strings to windows
    for (int i = 1; i < thisMenu.subItem.size(); i++)
      wprintw(thisMenu.window[i],thisMenu.subItem[i].c_str());
      //wprintw(subMenu[number].window[i],thisMenu.subItem[i].c_str());
      //wprintw(subMenu[number].window[i],"hello");
    
    wbkgd(thisMenu.window[1],COLOR_PAIR(1));
    wrefresh(thisMenu.window[0]);
    move(ty,tx);

    
  } // end drawSubMenu()
  
  void menuHeader::deleteSubMenu(int number){
    for (int i=0;i<subMenu[number].window.size();i++)
      delwin(subMenu[number].window[i]);

    subMenu[number].window.clear(); // remove all elements from vec of wins
  } // end deleteSubMenu()
  
  string menuHeader::scrollSubMenu(int number){
    int key = 0;
    int count = 0;
    int selected = 0;
    string selection = "";
    int nextItem = 0;
    // int selected = 0;
    // int nextItem;
    dropDownMenu& thisMenu = subMenu[number];
    vector<WINDOW*>& items = thisMenu.window;
    count = thisMenu.subItem.size()-1;
    //dropDownMenu*& thisMenu = menu.subMenu[subMenuItem];
    //dropDownMenu* thisMenu = &menu.subMenu[subMenuItem];
    //dropDownMenu& thisMenu = *menu.subMenu[subMenuItem];

    //vector<WINDOW*>& items = thisMenu->window;

    //printf("%d",count);
    while (1) {
      key=getch();
      if (key==KEY_DOWN || key==KEY_UP) {
	wbkgd(items[selected+1],COLOR_PAIR(2));
	wnoutrefresh(items[selected+1]);
	if (key==KEY_DOWN) {
	  // selected=(selected+1) % count;
	  //selected=mod((selected+1), count);
	  selected=mod((selected+1), count);
	  //selected = (selected+1) % thisMenu.subItem.size();
	  //selected = mod((selected+1), thisMenu->subItem.size());
	  // selected = mod((selected+1), thisMenu.subItem.size());
	  //selected++;
	} 
	else {
	  selected=mod((selected-1), count);
	  // selected=(selected+count-1) % count;
	  //selected = mod((selected-1), thisMenu.subItem.size());
	  //selected = mod((selected-1), thisMenu->subItem.size());
	  //selected--;
	}
	wbkgd(items[selected+1],COLOR_PAIR(1));
	wnoutrefresh(items[selected+1]);
	doupdate();
      } 
      else if ((key==KEY_LEFT) || (key==KEY_RIGHT)) {
	//delete_menu(thisMenu,count+1);
	deleteSubMenu(number);
	touchwin(stdscr);
	refresh();
	if (key==KEY_RIGHT){
	  nextItem = mod(number + 1 , subMenu.size());
	  //number = nextItem;
	}
	else{
	  nextItem = mod(number - 1 , subMenu.size());
	  //number = nextItem;
	}
	//thisMenu = subMenu[number];
	//items = thisMenu.window;
	//count = thisMenu.subItem.size()-1;
	drawSubMenu(nextItem);
	// return scroll_menu(thisMenu,8,20-menu_start_col);
	return scrollSubMenu(nextItem);
      } 
      else if (key==ESCAPE) {
	// return -1;
	return "Escape";
      } 
      else if (key==ENTER) {
	// return selected;
	// search subMenus for currentMenu name?
	return subMenu[number].subItem[selected+1];
      }
    }  // end key check
  } // end scrollSubMenu()
  
//}; // end menuHeader class


  // void init_curses(){
  //   initscr();
  //   start_color();
  //   /*
  //     init_pair(1,COLOR_WHITE,COLOR_BLUE);
  //     init_pair(2,COLOR_BLUE,COLOR_WHITE);
  //     init_pair(3,COLOR_RED,COLOR_WHITE);
  //   */

  //   init_pair(1,COLOR_GREEN,COLOR_BLACK);
  //   init_pair(2,COLOR_BLACK,COLOR_GREEN);
  //   init_pair(3,COLOR_BLUE,COLOR_GREEN);
    
  //   curs_set(0);
  //   noecho();
  //   keypad(stdscr,TRUE);
    
  // } // end init_curses







// int main()
// {
//   init_curses();
//   menuHeader startItUp;
//   startItUp.getKeys();
//   endwin();
//   return 0;
// }
